### PR TITLE
Updated TE+MaxText integration

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/llama2.py [hunk 1]
```diff
@@ -124,20 +124,9 @@ def __call__(
     attention_lnx = nn.with_logical_constraint(attention_lnx, activation_axis_names)
     intermediate_inputs = inputs + attention_lnx
```
Comment (jberchtold-nvidia) [inline MaxText/layers/llama2.py:129]: Changes in this file are a result of partially cherry-picking this PR for testing. https://github.com/AI-Hypercomputer/maxtext/pull/2101

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -391,12 +397,21 @@ def __init__(
...
         matmul_precision=self.config.matmul_precision,
         rngs=rngs,
     )
 
+    # Initialization must occur after the model params have been created for self.wi and self.wo
+    if self.te_ln_mlp is not None:
+      # Placeholder inputs, actual values including sharding axes don't matter as the modules
+      # do not initialize any parameters. The modules only initialize variables for recipe state (e.g. amax_history + scale for Delayed Scaling)
+      dummy_dot_1_input_axes = ("activation_batch", "activation_norm_length", "activation_embed")
+      dummy_inputs = jnp.zeros((1, 32, in_features), dtype=self.dtype)
```
Comment (jberchtold-nvidia) [inline MaxText/layers/linears.py]: Note: shape changed to 32 here to support MXFP8. Will try to hide this behind a config to prevent the being 32 when not required

File: MaxText/layers/linears.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def __init__(
       quant_dot_general = nnx_wrappers.ToNNX(dot_general_linen, rngs=rngs)
       self._quant_dot_general_name = f"{type(dot_general_linen).__name__}_0"
       setattr(self, self._quant_dot_general_name, quant_dot_general)
-      dummy_inputs = jnp.zeros((1, *self.in_features_shape), dtype=self.dtype)
+      dummy_inputs = jnp.zeros((32, *self.in_features_shape), dtype=self.dtype)
```
Comment (jberchtold-nvidia) [inline MaxText/layers/linears.py]: Note: shape changed to 32 here to support MXFP8. Will try to hide this behind a config to prevent the being 32 when not required

File: MaxText/layers/quantizations.py [hunk 4]
```diff
@@ -674,3 +678,151 @@ def maybe_quantize_model(model, config):
...
+    fp8_recipe = self._recipe
+
+    class TEWrapper(te.flax.module.TransformerEngineBase):
+      def generate_quantizer_set(self, postfix: str = ""):
+        OVERWRITE_WITH_GRADIENT = "_overwrite_with_gradient"
+        return super().generate_quantizer_set(postfix=postfix, variable_collection=OVERWRITE_WITH_GRADIENT, fp8_recipe=fp8_recipe)
+
+      @nn.compact
+      def __call__(self, *args, **kwargs):
+        with global_shard_guard(mesh_resource):
```
Comment (jberchtold-nvidia) [inline MaxText/layers/quantizations.py:760]: Was able to avoid `fp8_autocast` here by specifying the recipe directly when creating the quantization set, corresponding  [PR](https://github.com/NVIDIA/TransformerEngine/pull/2059) to TE to support this

However, we still need the `global_shard_guard` (which is typically called inside `fp8_autocast`) to correctly set the mesh resource

File: MaxText/layers/normalizations.py [hunk 5]
```diff
@@ -55,6 +55,20 @@ def __init__(
 
   def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
     """Applies layer normalization on the input."""
+    use_te_norm = True # TODO: get access to config to determine if TE norm should be enabled
+    if use_te_norm:
```
Comment (jberchtold-nvidia) [inline MaxText/layers/normalizations.py:60]: `rms_norm` in MaxText is called without the `config` so it's not as easy to toggle TE's norm here.

I'm thinking we have a few possible options here. These options only affect the pre-attention norm, the LNMLP norm is addressed already by using TE's `layernorm_mlp` call (assuming `fused_mlp` is enabled)

1. Provide the config as an additional argument. This would allow us to introduce new config options for toggling TE norm, but would require updating every callsite of `rms_norm` so would make this harder to rebase in future.
2. Do not integrate TE norm here. JAX norm is on-par or faster in our cases so far, so we can use MaxText's impl.
3. Integrate TE norm here and control it's enablement via an environment variable.

I'm leaning towards 2 or 3, more so 2 as I think that will be easier to rebase and upstream. But lmk what you think
